spynnaker/pyNN/models/abstract_models/population_application_vertex.py

```diff
@@ -0,0 +1,328 @@
+from pacman.model.graphs.application import ApplicationVertex
+from enum import Enum
+from spinn_utilities.helpful_functions import is_singleton
+from spinn_front_end_common.utilities.exceptions import ConfigurationException
+
+
+class RecordingType(Enum):
+    """ The type of data being recorded
+    """
+    BIT_FIELD = 0
+    MATRIX = 1
+    EVENT = 2
+
+
+class PopulationApplicationVertex(ApplicationVertex):
+    """ A vertex that can be used in a Population.
+
+        Provides some default functions that can be overridden if the vertex
+        supports these.
+    """
+
+    # No data required here; makes it easy to mix in!
+    __slots__ = []
+
+    @staticmethod
+    def _as_list(names):
+        """ Normalise the input to a list
+
+        :param names: The item or items to normalise
+        :type names: str or list
+        :rtype: list(str)
+        """
+        if is_singleton(names):
+            return [names]
+        return names
+
+    @staticmethod
+    def _check_names(names, allowed, type_of_thing):
+        """ Check the list of names are allowed or not
+
+        :param names: The names to check
+        :type names: str or list
+        :param set(str) allowed: The set of allowed names
+        :param str type_of_thing: The "type of thing" to report in any error
+        :raises KeyError: If one of the names is not allowed
+        """
+        for name in PopulationApplicationVertex._as_list(names):
+            if name not in allowed:
+                raise KeyError(f"{name} is not a recognized {type_of_thing}")
+
+    @staticmethod
+    def _check_parameters(names, allowed):
+        """ Check that parameters are allowed
+
+        :param names: The names to check
+        :type names: str or list
+        :param set(str) allowed: The set of allowed names
+        """
+        PopulationApplicationVertex._check_names(names, allowed, "parameter")
+
+    @staticmethod
+    def _check_variables(names, allowed):
+        """ Check that state variables are allowed
+
+        :param names: The names to check
+        :type names: str or list
+        :param set(str) allowed: The set of allowed names
+        """
+        PopulationApplicationVertex._check_names(names, allowed, "variable")
+
+    def get_parameter_values(self, names, selector=None):
+        """ Get the values of a parameter or parameters for the whole
+            Population or a subset if the selector is used
+
+        :param names: The name or names of the parameter to get
+        :type names: str or list
+        :param selector: a description of the subrange to accept, or ``None``
+            for all. See:
+            :py:meth:`~spinn_utilities.ranged.AbstractSized.selector_to_ids`
+        :type selector: None or slice or int or list(bool) or list(int)
+        :rtype: ParameterHolder
+        :raise KeyError: if the parameter is not something that can be read
+        """
+        raise KeyError(
```

Comment:
Are these not better as NotImplementedException

Reply:
The idea here is to make PyNN report that the parameter (or whatever) is not supported, rather than that the vertex didn't implement something.  The thought is that all vertices must extend this, so they automatically get support for this.  Then the vertex can override those things it does actually support leaving the default implementation to handle what happens in the normal case.  This has the advantage that the overridden method can call this one when it doesn't support the parameter in question.